File: Leetcode/October/Day7_Rotate_List.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution
{
public:
    ListNode *rotateRight(ListNode *head, int k)
    {
        ListNode *temp = head;
        int count = 0;
        if (temp == NULL)
            return head;
        while (temp->next != NULL)
        {
            count += 1;
            temp = temp->next;
        }
        temp->next = head;
        temp = temp->next;
        // cout<<temp->val<<endl;
        ListNode *prev = NULL;
        int remaining = count - (k % (count + 1)) + 1;
        while (remaining--)
        {
            prev = temp;
            temp = temp->next;
        }
        prev->next = NULL;
        return temp;
    }
};

File: Leetcode/October_LeetCoding_Challenge/Day7_Rotate_List/Day7_Rotate_List.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    public ListNode rotateRight(ListNode head, int k) {
        if(head == null || head.next == null)
            return head;
   
        int n = 0;
        ListNode temp = head;
        while(temp != null){
            n++;
            temp = temp.next;
        }
         if(k == n){
            return head;
        }
        if(k > n)
        {
            k = k%n;
        }
        if(k == 0)
            return head;
        
        int remains = n - k;
        temp = head;
        while(--remains > 0){
            temp = temp.next;
        }
        ListNode sec = temp.next;
        ListNode result = sec;
        temp.next = null;
        while(sec.next != null){
            sec = sec.next;
        }
        sec.next = head;
        return result;
    }
}
